File: financial-domain-starter/src/main/java/com/nixsolutions/financial/discovery/ServiceRegistry.java
package com.nixsolutions.financial.discovery;

import static com.nixsolutions.financial.discovery.DiscoveryType.LOCAL;
import lombok.Getter;
import lombok.Setter;

@Getter
@Setter
public class ServiceRegistry
{

  public static final String COLON = ":";

  public static class Services
  {
    public static final String STORAGE = "storage";
    public static final String UI = "ui";
    public static final String PREDICTION = "prediction";
    public static final String USER_AUTH = "userAuth";
  }

  private static final String HTTP_SCHEME = "http://";
  private static final String LOCALHOST = "localhost";
  private ServiceDiscoveryProperties serviceDiscoveryProperties;

  public String getServiceUrl(String svcName)
  {
    return serviceDiscoveryProperties.getList().stream()
        .filter(svc -> svc.getName().equals(svcName))
        .findAny()
        .map(this::serviceRecordToUrl)
        .orElseThrow(DiscoveryException::new);
  }

  private String serviceRecordToUrl(ServiceRecord serviceRecord)
  {
    DiscoveryType resolution = serviceRecord.getResolution();
    String port = serviceRecord.getPort();

    if (resolution.equals(LOCAL))
    {
      return HTTP_SCHEME + LOCALHOST + COLON + port;
    }
    else
    {
      return HTTP_SCHEME + serviceRecord.getName() + COLON + port;
    }
  }
}

File: financial-domain-starter/src/main/kotlin/com/nixsolutions/financial/discovery/ServiceRegistry.kt
package com.nixsolutions.financial.discovery

import com.nixsolutions.financial.discovery.DiscoveryType.LOCAL

const val STORAGE = "storage-service"
const val UI = "ui"
const val PREDICTION = "prediction"
const val USER_AUTH = "authorization-server"
const val USERS = "user-service"

class ServiceRegistry {
  var serviceDiscoveryProperties: ServiceDiscoveryProperties? = null

  fun getServiceUrl(svcName: String): String {
    return this.serviceDiscoveryProperties!!.list.stream()
        .filter { svc -> svc.name == svcName }
        .findAny()
        .map { serviceRecordToUrl(it) }
        .orElseThrow<DiscoveryException> { DiscoveryException() }
  }

  private fun serviceRecordToUrl(serviceRecord: ServiceRecord): String {
    val resolution = serviceRecord.resolution
    val port = serviceRecord.port

    return if (resolution == LOCAL) {
      HTTP_SCHEME + LOCALHOST + COLON + port
    } else {
      HTTP_SCHEME + serviceRecord.name + COLON + port
    }
  }

  private val COLON = ":"
    private val HTTP_SCHEME = "http://"
    private val LOCALHOST = "localhost"
}
